### ISSUE TITLE
Hang in CommandManager due to async update of keybinding

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#4917]
We have the following:
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs#L77-L100
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs#L505-L507

We have a possible issue with the code, given an `UpdateAsync` which looks like:

```
protected override async Task UpdateAsync (CommandArrayInfo info, CancellationToken cancelToken)
{
    // Do some UI work
    UpdateMyLabels();

    // case 1 at this point, we return the task
    await Task.Run (() => SomeBgWork());

    // it can either be configureawait(false) and runtime.runinmainthread or configureawait(true)

    // Therefore, at this point, we need to run on the UI thread, but that is blocked by CommandManager.ProcessKeyEvent which holds the UI thread
}
```

when bg work is started, we return a task, there is a window to race, that is pretty big, depending on how much bg code takes, because we almost instantly go to a synchronous wait on keypress.

~CoreCLR (implicitly, mono 2018-02+) have a smart mechanism through which if a given task has no timeout/cancellation, it'll inline the task to run on the given thread: http://source.dot.net/#System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,2910~

~Alas, since the cancellation token _is_ bound to a cancellation token source, that cannot save us, since we always pass in a cancellation token bound to a source:
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs#L79
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandHandler.cs#L91
https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs#L2624
http://source.monodevelop.com/#MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs,2678~

EDIT: it seems it won't if it's already in running state, which is the case here.

I guess we should either wait a finite time ~or somehow figure out a way to inline the task on the UI thread~. Otherwise, we'll hang. I could reproduce this with go to declaration, but not sure how, as the race window there is _incredibly small_
https://github.com/mono/monodevelop/blob/master/main/src/addins/CSharpBinding/MonoDevelop.CSharp.Refactoring/CSharpFeaturesTextEditorExtension.cs#L125

> VS bug [#623224](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/623224)

**Comment by Therzok** [comment]
Moved this to a future sprint and assigned to myself.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-34] Comment by **xamarinc**



Fixed in version 8.3.0.1536 (release-8.3)



Pull Request #8570 merged by: Vsevolod Kukol  
 Author: mono  
 Commit: 71005e45ecbf0dbbfc13f911a8d57b5b8563db27 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-36] Comment by **xamarinc**



Fixed in version 8.5.0.18 (master)



Pull Request #8176 merged by: Vsevolod Kukol  
 Author: mono  
 Commit: 02a13743043f2bb84928de116b2be160eb2c1a62 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-38] Comment by **xamarinc**



Fixed in version 8.5.0.19 (master)



Pull Request #5312 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 794fd6871f2e72f1c77fb2893a79e40dc7c71a0d (xamarin/md-addins)  
 Included in Commit: 8060d05b6b8f991bafebbc4334b5ab36ac511d00 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-39] Comment by **xamarinc**



Fixed in version 8.3.0.1537 (release-8.3)



Pull Request #5313 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 69e6e2dc5f49e7303c1a16cb8e1969f75a8faa9e (xamarin/md-addins)  
 Included in Commit: 4241f7820fd25547ffe8acd073156096182ec44b (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-40] Comment by **xamarinc**



Fixed in version 8.5.0.162 (release-8.4)



Pull Request #5312 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 794fd6871f2e72f1c77fb2893a79e40dc7c71a0d (xamarin/md-addins)  
 Included in Commit: d7a60d5e7a39a2ca42a5e831e4cfa6fa47337408 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-42] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #8176 merged by: Vsevolod Kukol  
 Author: mono  
 Commit: 02a13743043f2bb84928de116b2be160eb2c1a62 (xamarin/monodevelop)
